conversation_simulator/src/StateMachine.js

```diff
@@ -162,14 +163,17 @@ export default class {
   scoreSolution (solution, score, hintUsed) {
     if (solution !== undefined) {
       let nextState = this.currentState.Solutions[solution].Next
+      this.submitSolutionResult = true
+      if (score > 0 && !hintUsed && !this.checkIfSolutionUsed(this.currentState, solution)) {
+        this.score += score
+      }
+      // Save state/answer so it can't be used again for points
+      this.markSolution(this.currentState, solution)
```

Comment:
let's make it generic, and encapsulate all context properties under on object, so we can expand it in future:

this.userContext = {}
this.userContext.visitedStates = {}
this.userConexts.addSolution=function() {...}
this.userConexts.IsSolutionUsedBefore=function() {...}
...

Let's also move it into a separate JS file, our code is getting too complex and we need to start applying good patterns to keep it clean

Reply:
UserContext is now a separate class